correctly send secure cookies after hsts url match

bobthekingofegypt opened the pull request:
Fixes #8100, where sites in the hsts list were not recieving secure
cookies if the site was originally loading using a plain http url.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9780)

<!-- Reviewable:end -->

bobthekingofegypt committed:
correctly send secure cookies after hsts url match

Fix for #8100, where sites in the hsts list were not recieving secure
cookies if the site was originally loading using a plain http url.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

jdm commented:
Thank you for finding and solving this! I've filed #9783 to ensure we're not making this mistake in other places, too.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 759099c has been approved by `jdm`

<!-- @bors-servo r=jdm 759099c78dcd92fc1916d6fdc3d0a8305cadb361 -->

bors-servo commented:
:hourglass: Testing commit 759099c with merge 32bed54...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1958)

jdm commented:
```
  ▶ Unexpected subtest result in /XMLHttpRequest/anonymous-mode-unsupported.htm:
  │ FAIL [expected PASS] XMLHttpRequest: anonymous mode unsupported
  │   → assert_equals: The deprecated anonymous:true should be ignored, cookie sent anyway expected "cookie: test=anonymous-mode-unsupported\n" but got ""
  │ 
  │ client.onreadystatechange<@http://web-platform.test:8000/XMLHttpRequest/anonymous-mode-unsupported.htm:22:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1405:1
```
We should look into why this test now fails with this change.

bobthekingofegypt committed:
refactor http_loader hostname/htst order

Changed hostname rewrite to happen inside obtain response after any htst
changes.  Removed  url from load leaving just doc_url to avoid confusion

bobthekingofegypt commented:
As discussed on IRC, the test was failing due to the hostname being modified by the HOST file entry in the test runner.  This meant the cookies were incorrectly failing to match the host in the updated request cookies method, and not being sent.

Tried updating the code to move hostname changes inside obtain_response and remove the duplicate urls.

Ran this against the unit tests and WPT tests, I get some failures and timeouts on WPT but I get the same failures on master.  I tried to test github site itself but I can't seem to press the final login button as the browser won't react to it.

jdm commented on components/net/http_loader.rs:
```diff
@@ -513,14 +513,14 @@ fn request_must_be_secured(url: &Url, hsts_list: &Arc<RwLock<HSTSList>>) -> bool
                               load_data: &LoadData) {
     // Ensure that the host header is set from the original url
     let host = Host {
-        hostname: doc_url.serialize_host().unwrap(),
```
This change doesn't respect the comment above it; we need to pass the original URL as well.

jdm replied:
Nevermind, I hadn't finished reading the rest of the changes.

jdm commented on components/net/http_loader.rs:
```diff
@@ -632,7 +632,7 @@ pub fn obtain_response<A>(request_factory: &HttpRequestFactory<R=A>,
             send_request_to_devtools(
                 devtools_chan.clone(), request_id.clone().into(),
-                url.clone(), method.clone(), request_headers.clone(),
+                connection_url.clone(), method.clone(), request_headers.clone(),
```
This should be using `url` instead.

jdm commented:
This looks great, with one small change necessary!

jdm commented on tests/unit/net/http_loader.rs:
```diff
@@ -814,6 +814,50 @@ fn test_load_sets_requests_cookies_header_for_url_by_getting_cookies_from_the_re
+        let cookie = Cookie::new_wrapped(
+            cookie_pair,
+            &cookie_url,
+            CookieSource::NonHTTP
```
Let's make this HTTP as well, or else it's pretty confusing to read.

bobthekingofegypt committed:
send correct url to devtool on request

Send url that was not modified by the hosts file to the dev tools

bobthekingofegypt committed:
create cookie as HTTP instead of NonHttp in test

Cleanup for readability and correctness of unit test in http_loader

jdm commented:
@bors-servo: r+
Thanks!

bors-servo commented:
:pushpin: Commit 248e84e has been approved by `jdm`

<!-- @bors-servo r=jdm 248e84e9281f464b438ef36b9c36de467dfb13a8 -->

bors-servo commented:
:hourglass: Testing commit 248e84e with merge 162e89d...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1953), [gonk](http://build.servo.org/builders/gonk/builds/2000), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1976), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2256), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/675), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2028), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2082), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.185)
